linera-core/src/worker.rs

```diff
@@ -1091,34 +1060,59 @@ where
                 recipient,
                 certificates,
             } => {
-                let mut actions = NetworkActions::default();
+                let mut latest_heights = Vec::<(Medium, BlockHeight)>::new();
                 for (medium, heights) in height_map {
-                    let origin = Origin { sender, medium };
+                    let origin = Origin {
+                        sender,
+                        medium: medium.clone(),
+                    };
                     let app_certificates = certificates
                         .iter()
                         .filter(|cert| heights.binary_search(&cert.value.block().height).is_ok())
                         .cloned()
                         .collect();
-                    actions.merge(
-                        self.process_cross_chain_update(
-                            origin.clone(),
-                            recipient,
-                            app_certificates,
-                        )
-                        .await?,
-                    );
+                    if let Some(height) = self
+                        .process_cross_chain_update(origin.clone(), recipient, app_certificates)
```

Comment:
nit: I would keep `let origin = Origin { sender, medium };` above: since you clone the origin here, then you can write `origin.medium.clone()` below.